engines/e_pkcs11.c

```diff
@@ -0,0 +1,841 @@
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjects failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_FAILED);
+        goto err;
+    }
+
+    rv = pkcs11_funcs->C_FindObjectsFinal(ctx->session);
+
+    if (rv != CKR_OK) {
+        PKCS11_trace("C_FindObjectsFinal failed, error: %#08X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_FINAL_FAILED);
+        goto err;
+    }
+
+
+    ctx->key = objhandle;
+    return 1;
+
+ err:
+    return 0;
+}
+
+static char pkcs11_hex_int(char nib1, char nib2)
+{
+    int ret = (nib1-(nib1 <= 57 ? 48 : (nib1 < 97 ? 55 : 87)))*16;
+    ret += (nib2-(nib2 <= 57 ? 48 : (nib2 < 97 ? 55 : 87)));
+    return ret;
+}
+
+static int pkcs11_parse_uri(const char *path, char *token, char **value)
+{
+    char *tmp, *end, *hex2bin;
+    size_t vlen, i, j = 0, tmplen;
+
+    if ((tmp = strstr(path, token)) == NULL)
+        return 0;
+    tmp += strlen(token);
+    tmplen = strlen(tmp);
+    *value = OPENSSL_malloc(tmplen + 1);
+
+    if (*value == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_PARSE_URI, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    end = strpbrk(tmp, ";");
+
+    BIO_snprintf(*value, end == NULL ? tmplen + 1 :
+             (size_t) (end - tmp + 1), "%s", tmp);
+    hex2bin = OPENSSL_malloc(strlen(*value) + 1);
+
+    if (hex2bin == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_PARSE_URI, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    vlen = strlen(*value);
+    for (i = 0; i < vlen; i++) {
+        if (*(*value+i) == '%' && i < (strlen(*value)-2)) {
+            *(hex2bin+j) = pkcs11_hex_int(*(*value+i+1), *(*value+i+2));
+            i += 2;
+        } else {
+            *(hex2bin+j) = *(*value+i);
+        }
+        j++;
+    }
+    *(hex2bin+j) = '\0';
+    *value = hex2bin;
+    return 1;
+
+ err:
+    return 0;
+}
+
+static int pkcs11_get_console_pin(char **pin)
+{
+#ifndef OPENSSL_NO_UI_CONSOLE
+    int i;
+    char *strbuf = OPENSSL_malloc(512);
+
+    if (strbuf == NULL) {
+        PKCS11err(PKCS11_F_PKCS11_GET_CONSOLE_PIN, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    for (;;) {
+        char prompt[200];
+        BIO_snprintf(prompt, sizeof(prompt), "Enter PIN: ");
+        strbuf[0] = '\0';
+        i = EVP_read_pw_string((char *)strbuf, 512, prompt, 1);
+        if (i == 0) {
+            if (strbuf[0] == '\0') {
+                goto err;
+            }
+            *pin = strbuf;
+            return 1;
+        }
+        if (i < 0) {
+            PKCS11_trace("bad password read\n");
+            goto err;
+        }
+    }
+
+ err:
+    OPENSSL_free(strbuf);
+#endif
+
+    return 0;
+}
+
+static int pkcs11_parse(PKCS11_CTX *ctx, const char *path)
+{
+    char *id, *module_path = NULL;
+    char *pin = NULL, *label = NULL;
+    char *slotid = OPENSSL_malloc(2);
```

Comment:
useless allocation, just set `ctx->slotid = 0;` to begin this method.
And call to atoi() after, if `slotid != NULL.`